sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiSinkDefaultValuesIT.java

```diff
@@ -0,0 +1,280 @@
+    BQ_CLIENT.deleteDataset(PROJECT, BIG_QUERY_DATASET_ID);
+  }
+
+  private static String createAndGetTablespec(TableSchema tableSchema)
+      throws IOException, InterruptedException {
+    String tableName = "table" + System.nanoTime();
+    TableReference tableReference =
+        new TableReference()
+            .setProjectId(PROJECT)
+            .setDatasetId(BIG_QUERY_DATASET_ID)
+            .setTableId(tableName);
+    BQ_CLIENT.createNewTable(
+        PROJECT,
+        BIG_QUERY_DATASET_ID,
+        new Table().setSchema(tableSchema).setTableReference(tableReference));
+    return PROJECT + "." + BIG_QUERY_DATASET_ID + "." + tableName;
+  }
+
+  @Test
+  public void testMissingValueSchemaKnownTakeDefault() throws IOException, InterruptedException {
+    runTest(true, true, false);
+  }
+
+  @Test
+  public void testMissingRequiredValueSchemaKnownTakeDefault()
+      throws IOException, InterruptedException {
+    runTest(true, true, true);
+  }
+
+  @Test
+  public void testMissingRequiredValueSchemaKnownTakeNull()
+      throws IOException, InterruptedException {
+    runTest(true, false, true);
+  }
+
+  @Test
+  public void testMissingRequiredValueSchemaUnknownTakeDefault()
+      throws IOException, InterruptedException {
+    runTest(false, true, true);
+  }
+
+  @Test
+  public void testMissingValueSchemaUnknownTakeDefault() throws IOException, InterruptedException {
+
+    runTest(false, true, false);
+  }
+
+  @Test
+  public void testMissingValueSchemaKnownTakeNull() throws IOException, InterruptedException {
+    runTest(true, false, false);
+  }
+
+  @Test
+  @Ignore // This currently appears broke in BigQuery.
+  public void testMissingValueSchemaUnknownTakeNull() throws IOException, InterruptedException {
+    runTest(false, false, false);
+  }
+
+  public void runTest(
+      boolean sinkKnowsDefaultFields, boolean takeDefault, boolean defaultFieldsRequired)
+      throws IOException, InterruptedException {
+    boolean expectDeadLetter = !takeDefault && defaultFieldsRequired;
+    TableSchema bqSchema;
+    if (defaultFieldsRequired) {
+      bqSchema =
+          new TableSchema()
+              .setFields(
+                  ImmutableList.of(
+                      new TableFieldSchema().setName("id").setType("STRING"),
+                      new TableFieldSchema().setName("key2").setType("STRING"),
+                      new TableFieldSchema().setName("value").setType("STRING"),
+                      new TableFieldSchema()
+                          .setName("defaultliteral")
+                          .setType("INT64")
+                          .setDefaultValueExpression("42")
+                          .setMode("REQUIRED"),
+                      new TableFieldSchema()
+                          .setName("defaulttime")
+                          .setType("TIMESTAMP")
+                          .setDefaultValueExpression("CURRENT_TIMESTAMP()")
+                          .setMode("REQUIRED")));
```

Comment:
Can we add a REPEATED field with a default? I worry that changing the mode to NULLABLE in `modeToProtoMode` may cause some weird things to happen

Reply:
done